nestkernel/source_table.h

```diff
@@ -471,27 +475,77 @@ SourceTable::no_targets_to_process( const size_t tid )
 }
 
 inline size_t
-SourceTable::find_first_source( const size_t tid, const synindex syn_id, const size_t snode_id ) const
+SourceTable::find_first_source( const size_t tid,
+  const synindex syn_id,
+  const size_t snode_id,
+  bool isCompressedEnabled /* default = false */ ) const
 {
-  // binary search in sorted sources
-  const BlockVector< Source >::const_iterator begin = sources_[ tid ][ syn_id ].begin();
-  const BlockVector< Source >::const_iterator end = sources_[ tid ][ syn_id ].end();
-  BlockVector< Source >::const_iterator it = std::lower_bound( begin, end, Source( snode_id, true ) );
-
-  // source found by binary search could be disabled, iterate through
-  // sources until a valid one is found
-  while ( it != end )
+  using SourceIter = BlockVector< Source >::const_iterator;
+
+  const SourceIter begin = sources_[ tid ][ syn_id ].begin();
+  const SourceIter end = sources_[ tid ][ syn_id ].end();
+
+  Source value { snode_id, true };
+
+  if ( isCompressedEnabled )
   {
-    if ( it->get_node_id() == snode_id and not it->is_disabled() )
+    // binary search in sorted sources
+    SourceIter it = std::lower_bound( begin, end, value );
+
+    // source found by binary search could be disabled, iterate through
+    // sources until a valid one is found
+    while ( it != end )
     {
-      const size_t lcid = it - begin;
-      return lcid;
+      if ( it->get_node_id() == snode_id and not it->is_disabled() )
+      {
+        const size_t lcid = it - begin;
+        return lcid;
+      }
+      ++it;
     }
-    ++it;
+
+    // no enabled entry with this snode ID found
+    return invalid_index;
   }
+  else
+  {
+
+    auto nth_equal =
+      []( SourceIter first, SourceIter last, const Source& value, size_t n ) -> std::pair< bool, SourceIter >
+    {
+      if ( n == 0 )
+      {
+        auto iter = std::find( first, last, value );
+        return { iter != last, iter };
+      }
+      auto iter = std::find( first, last, value );
+      while ( n > 0 && iter != last )
+      {
+        --n;
+        iter = std::find( std::next( iter ), last, value );
+      }
+      return { iter != last, iter };
+    };
+    size_t pos = 0;
+    auto res = nth_equal( begin, end, value, pos );
+    if ( !res.first )
```

Comment:
According to NEST coding standards, please write out `not`, `and`, `or`. A bit surprised that our CI setup didn't detect his, @terhorstd.